File: cpp/src/p10_2/main.cc
// sum primes below 2 million

#include <cstdint>
#include <iostream>
#include <thread>
#include <vector>
#include "utils.h"

static const int NUM_THREADS = std::thread::hardware_concurrency();

int main() {
  // setup
  std::vector<uint64_t> answers(NUM_THREADS, 0);

  // Run threads that handle a part of the problem
  std::vector<std::thread> threads;
  for (int id = 0; id < NUM_THREADS; ++id) {
    threads.emplace_back([id,&answers]() {
      uint64_t sum = 0;
      for (int i = 0+id; i < 2000000; i+=NUM_THREADS) {
        if (isPrime(i)) {
          sum += i;
        }
      }
      answers[id] = sum;
    });
  }

  // Wait for them to finish
  for (auto& t : threads) {
    t.join();
  }

  // Collect the answers and print
  uint64_t sum = 0;
  for (const auto v : answers) {
    sum += v;
  }

  std::cout << sum << "\n";
}

File: golang/src/p10_2/main.go
package main

import (
	"acommon"
	"fmt"
	"runtime"
	"sync"
)

func main() {
	// setup
	NumParallel := runtime.NumCPU()
	answers := make([]uint64, NumParallel)

	var wg sync.WaitGroup
	wg.Add(NumParallel)

	// Run the threads to divide up the work and find the sum
	for id := 0; id < NumParallel; id++ {
		go func(id int) {
			sum := uint64(0)
			for i := 1 + id; i < 2000000; i += NumParallel {
				if acommon.IsPrime(i) {
					sum += uint64(i)
				}
			}
			answers[id] = sum
			wg.Done()
		}(id)
	}

	// When they're all done, add them all up for the final sum
	wg.Wait()
	sum := uint64(0)
	for _, v := range answers {
		sum += v
	}

	fmt.Println(sum)
}
